File: taxcalculator/java/src/test/java/com/gropyus/interviews/taxcalculator/StateControllerTest.java
package com.gropyus.interviews.taxcalculator;

import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class StateControllerTest {
    @Autowired
    RestTemplateBuilder restTemplateBuilder;

    @LocalServerPort
    int port;

    @Test
    void shouldReturnSupportedStates() {
        ResponseEntity<StatesResponse> response = restTemplateBuilder.build()
                .getForEntity("http://localhost:" + port + "/states", StatesResponse.class);

        Assertions.assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        Assertions.assertThat(response.getBody().getStates()).contains("TX");
    }
}

File: taxcalculator/kotlin/src/test/kotlin/com/gropyus/interviews/taxcalculator/StateControllerTest.kt
package com.gropyus.interviews.taxcalculator

import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.web.client.RestTemplateBuilder
import org.springframework.boot.web.server.LocalServerPort
import org.springframework.http.HttpStatus

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
internal class StateControllerTest {
    @Autowired
    var restTemplateBuilder: RestTemplateBuilder? = null

    @LocalServerPort
    var port = 0

    @Test
    fun `should return supported states`() {
        val response = restTemplateBuilder!!.build()
            .getForEntity("http://localhost:$port/states", List::class.java)
        Assertions.assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        Assertions.assertThat(response.body).contains("TX")
    }

    @Test
    fun `should accept tax rate per state`() {
        val response = restTemplateBuilder!!.build()
            .postForEntity("http://localhost:$port/states/UT/tax", 0.0685, String::class.java)

        Assertions.assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
    }
}
